Disable preventive GC

Chaho12 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Preventive GC was added in 17, but has issues so JDK 20 disables it by default (plans to [completely remove it in JDK 21](https://bugs.openjdk.org/browse/JDK-8297639))
Users should be aware that until Trino uses java 20+, they should not use preventive GC as it could lead to unexpected OOM
- close https://github.com/trinodb/trino/issues/15379

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

- https://trinodb.slack.com/archives/C01TEP0HJTH/p1670817191607989

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

wendigo commented:
I think that we should add this to the default jvm.config files. 

cc @trinodb/maintainers

findepi commented:
> Preventive GC was added in 17

Is it enabled by default?

Chaho12 commented:
> Is it enabled by default?

Yes, so a pr was created to disable it by default. However, it is only applied to java 20 and not 17.
- https://bugs.openjdk.org/browse/JDK-8293861

findepi commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -143,10 +143,12 @@ The following provides a good starting point for creating ``etc/jvm.config``:
     -Djdk.nio.maxCachedBufferSize=2000000
     -XX:+UnlockDiagnosticVMOptions
     -XX:+UseAESCTRIntrinsics
+    -XX:-G1UsePreventiveGC
```
Can you please put it next to other GC-related properties, eg after `-XX:+ExplicitGCInvokesConcurrent`?

Also, please add a line of comment `# Disable preventive GC, see https://bugs.openjdk.org/browse/JDK-8293861 for details`

Also, please update jvm.config in RPM and product tests.

Chaho12 replied:
> Can you please put it next to other GC-related properties, eg after -XX:+ExplicitGCInvokesConcurrent?

Because `-XX:-G1UsePreventiveGC` is a diagnostic option, it must be called after `-XX:+UnlockDiagnosticVMOptions`. 

So how about placing it in between `-XX:MaxRAMPercentage=80` and `-XX:G1HeapRegionSize=32M` as shown below?
```
    -server
    -Xmx16G
    -XX:InitialRAMPercentage=80
    -XX:MaxRAMPercentage=80
    -XX:+UnlockDiagnosticVMOptions
    -XX:+UseAESCTRIntrinsics
    -XX:-G1UsePreventiveGC
    -XX:G1HeapRegionSize=32M
    -XX:+ExplicitGCInvokesConcurrent
    -XX:+ExitOnOutOfMemoryError
    -XX:+HeapDumpOnOutOfMemoryError
    -XX:-OmitStackTraceInFastThrow
    -XX:ReservedCodeCacheSize=512M
    -XX:PerMethodRecompilationCutoff=10000
    -XX:PerBytecodeRecompilationCutoff=10000
    -Djdk.attach.allowAttachSelf=true
    -Djdk.nio.maxCachedBufferSize=2000000
```

Chaho12 replied:
> Also, please update jvm.config in RPM and product tests.

When you mean by this sentence, do you mean all the files listed below? First line is from docker, second one is rpm, and last three is from testing/trino-product-tests-launcher.

![image](https://user-images.githubusercontent.com/25147023/207480402-ad09e658-912f-4012-936f-66504c822643.png)

findepi commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -143,10 +143,12 @@ The following provides a good starting point for creating ``etc/jvm.config``:
 inconsistent state, we write a heap dump, for debugging, and forcibly
-terminate the process when this occurs.
+terminate the process when this occurs. Also, we suggest disabling
+preventive GCs which cause OOME when the heap is almost full.
```
Note that this paragraph doesn't explain every single flag that exists above.
It would be too detailed. Let's not add it here.

mosabua replied:
agreed.

Chaho12 replied:
How about adding after this paragraph?

We enable -XX:+UnlockDiagnosticVMOptions and -XX:+UseAESCTRIntrinsics to improve AES performance for S3, etc. on ARM64 ([JDK-8271567](https://bugs.openjdk.java.net/browse/JDK-8271567))

something like this

`-XX:-G1UsePreventiveGC` is disabled along with `-XX:+UseAESCTRIntrinsics` option for performance issue (`JDK-8293861 <https://bugs.openjdk.org/browse/JDK-8293861>`)

findepi commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -131,6 +131,9 @@ The following provides a good starting point for creating ``etc/jvm.config``:
     -XX:InitialRAMPercentage=80
     -XX:MaxRAMPercentage=80
+    -XX:+UnlockDiagnosticVMOptions
+    -XX:+UseAESCTRIntrinsics
```
I would prefer to keep GC options together, and AES intrinsics isn't one of them.
I wasn't aware `-XX:+UnlockDiagnosticVMOptions` needs to come first.
Let's revert to your original order, where you simply added the new option as the last.

findepi commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -156,6 +157,7 @@ temporary directory by adding ``-Djava.io.tmpdir=/path/to/other/tmpdir`` to the
 list of JVM options.
 
 We enable ``-XX:+UnlockDiagnosticVMOptions`` and ``-XX:+UseAESCTRIntrinsics`` to improve AES performance for S3, etc. on ARM64 (`JDK-8271567 <https://bugs.openjdk.java.net/browse/JDK-8271567>`_)
+Also, disable ``-XX:-G1UsePreventiveGC`` for performance issue (`JDK-8293861 <https://bugs.openjdk.org/browse/JDK-8293861>`_)
```
```suggestion
We disable preventive GC (``-XX:-G1UsePreventiveGC``) for performance reasons (see `JDK-8293861 <https://bugs.openjdk.org/browse/JDK-8293861>`_)
```

findepi commented:
Please update the RPM config and product tests too.

Chaho12 commented:
> Please update the RPM config and product tests too.

I updated these jvm.configs files, is this what you meant by rpm config and product tests? I searched all usages for `-XX:+UnlockDiagnosticVMOptions` and added `-XX:-G1UsePreventiveGC` option.
![image](https://user-images.githubusercontent.com/25147023/207543572-24bea882-edf2-48f4-9fda-123ece2766c6.png)

findepi commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -143,6 +143,7 @@ The following provides a good starting point for creating ``etc/jvm.config``:
     -Djdk.nio.maxCachedBufferSize=2000000
     -XX:+UnlockDiagnosticVMOptions
     -XX:+UseAESCTRIntrinsics
+    -XX:-G1UsePreventiveGC
```
```suggestion
    # Prevent GC for performance reasons (JDK-8293861)
    -XX:-G1UsePreventiveGC
```

losipiuk commented on docs/src/main/sphinx/installation/deployment.rst:
```diff
@@ -156,6 +158,7 @@ temporary directory by adding ``-Djava.io.tmpdir=/path/to/other/tmpdir`` to the
 list of JVM options.
 
 We enable ``-XX:+UnlockDiagnosticVMOptions`` and ``-XX:+UseAESCTRIntrinsics`` to improve AES performance for S3, etc. on ARM64 (`JDK-8271567 <https://bugs.openjdk.java.net/browse/JDK-8271567>`_)
+We disable prevent GC (``-XX:-G1UsePreventiveGC``) for performance reasons (see `JDK-8293861 <https://bugs.openjdk.org/browse/JDK-8293861>`_)
```
nit: prevent -> preventive

findepi replied:
thanks @losipiuk for correcting my mistakes (https://github.com/trinodb/trino/pull/15380#discussion_r1048185956)

losipiuk reviewed (approved):
LGTM % nits

Chaho12 commented:
I don't quite understand why it failed for presto v350 version :(
Also, I removed preventive GC for parallel GC test (preventive GC only exists for G1GC)

![image](https://user-images.githubusercontent.com/25147023/207744983-7ac1efee-cfa0-4b49-9fe9-e164f02601ac.png)

wendigo commented:
@Chaho12 v350 is running on JDK11

Chaho12 commented:
> @Chaho12 v350 is running on JDK11

Oh i see. I removed `-XX:-G1UsePreventiveGC` from https://github.com/trinodb/trino/blob/master/testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/presto/etc/jvm.config

wendigo commented:
Thank you, @Chaho12, for contributing this!
